src/gwt/src/org/rstudio/core/client/PriorityTaskQueue.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * PriorityTaskQueue.java
+ *
+ * Copyright (C) 2020 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+
+package org.rstudio.core.client;
+
+import java.util.LinkedList;
+import java.util.Queue;
+
+import com.google.gwt.user.client.Command;
+
+// Task queue that allows tasks to bump themselves up in priority (even after
+// they have been added to the queue).
+
+public class PriorityTaskQueue
```

Comment:
For nomenclature I'd suggest a different name for this class since a "priority queue" is a well known data structure but doesn't have much in common with what's done here. Maybe `PreemptiveTaskQueue` since the active document can preempt the other queued tasks?

Reply:
Good point! Renamed to `PreemptiveTaskQueue`